src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/build_plan/JenkinsBuildPlanCreator.java

```diff
@@ -0,0 +1,157 @@
+
+    private static final String REPLACE_DEFAULT_BRANCH = "#defaultBranch";
+
+    private static final String REPLACE_CHECKOUT_SOLUTION = "#checkoutSolution";
+
+    private static final String REPLACE_BUILD_PLAN_URL = "#buildPlanUrl";
+
+    private String artemisNotificationUrl;
+
+    @Value("${artemis.continuous-integration.secret-push-token}")
+    private String pushToken;
+
+    @Value("${artemis.continuous-integration.vcs-credentials}")
+    private String gitCredentialsKey;
+
+    @Value("${artemis.continuous-integration.build-timeout:#{30}}")
+    private String buildTimeout;
+
+    @Value("${server.url}")
+    private String artemisServerUrl;
+
+    @Value("${artemis.continuous-integration.artemis-authentication-token-key}")
+    private String artemisAuthenticationTokenKey;
+
+    @Value("${artemis.version-control.default-branch:main}")
+    private String defaultBranch;
+
+    private final ResourceLoaderService resourceLoaderService;
+
+    public JenkinsBuildPlanCreator(ResourceLoaderService resourceLoaderService) {
+        this.resourceLoaderService = resourceLoaderService;
+    }
+
+    @PostConstruct
+    public void init() {
+        this.artemisNotificationUrl = artemisServerUrl + Constants.NEW_RESULT_RESOURCE_API_PATH;
+    }
+
+    @Override
+    public Document buildBasicConfig(ProgrammingLanguage programmingLanguage, Optional<ProjectType> projectType, InternalVcsRepositoryURLs internalVcsRepositoryURLs,
+            boolean checkoutSolution, String buildPlanUrl) {
+        final String jenkinsfile = getJenkinsfile(internalVcsRepositoryURLs, checkoutSolution, buildPlanUrl);
+
+        final Path configFilePath = Path.of("templates", "jenkins", "config.xml");
+        final var configFileReplacements = Map.of(REPLACE_PIPELINE_SCRIPT, jenkinsfile, REPLACE_PUSH_TOKEN, pushToken);
+        final var xmlResource = resourceLoaderService.getResource(configFilePath.toString());
+        return XmlFileUtils.readXmlFile(xmlResource, configFileReplacements);
+    }
+
+    private String getJenkinsfile(InternalVcsRepositoryURLs internalVcsRepositoryURLs, boolean checkoutSolution, String buildPlanUrl) {
+        final String jenkinsfile = makeSafeForXml(loadJenkinsfile());
+        final var replacements = getReplacements(internalVcsRepositoryURLs, checkoutSolution, buildPlanUrl);
+
+        return replacePipelineScriptParameters(jenkinsfile, replacements);
+    }
+
+    private String loadJenkinsfile() {
+        Resource resource = resourceLoaderService.getResource("templates", "jenkins", "Jenkinsfile");
+
+        try (InputStream inputStream = resource.getInputStream()) {
+            return IOUtils.toString(inputStream, Charset.defaultCharset());
+        }
+        catch (IOException e) {
+            throw new ContinuousIntegrationBuildPlanException("Could not load Jenkinsfile.", e);
+        }
+    }
+
+    private String makeSafeForXml(final String script) {
```

Comment:
Can you please add a clarifying comment how this works? I think a link to the explanation is enough.

Reply:
Thanks for looking at the code. I added some javadoc in [ecdff30](https://github.com/ls1intum/Artemis/pull/6284/commits/ecdff300b13d3dd8537d649a3ac66f2319d569ba) to hopefully answer your question. Please tell me if it is still unclear.

Reply:
Thank you for this clarifying comment, now it is clear to me. Perhaps it would be best to rename that function to `escapeInXml` or something similar, to make the intention clear.